src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc

```diff
@@ -365,6 +365,86 @@ FileWatcherCertificateProvider::ReadIdentityKeyCertPairFromFiles(
   return absl::nullopt;
 }
 
+DataWatcherCertificateProvider::DataWatcherCertificateProvider(
+    std::string root_certificate,
+    grpc_core::PemKeyCertPairList pem_key_cert_pairs)
+    : StaticDataCertificateProvider(std::move(root_certificate), std::move(pem_key_cert_pairs)) {}
+
+absl::Status DataWatcherCertificateProvider::ReloadRootCertificate(
+    const std::string& root_certificate) {
+  if (root_certificate.empty()) {
+    return absl::InvalidArgumentError("Root Certificate string is empty.");
+  }
+  grpc_core::MutexLock lock(&mu_);
+  if (root_certificate_ == root_certificate) {
+    return absl::InvalidArgumentError("Root Certificate has not changed.");
+  }
+  root_certificate_ = root_certificate;
+  ExecCtx exec_ctx;
+  grpc_error_handle root_cert_error = GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+      "Unable to get latest root certificates.");
+  for (const auto& p : watcher_info_) {
+    const std::string& cert_name = p.first;
+    const WatcherInfo& info = p.second;
+    absl::optional<std::string> root_to_report;
+    if (info.root_being_watched) {
+      root_to_report = root_certificate_;
+    }
+    if (root_to_report.has_value()) {
+      distributor_->SetKeyMaterials(cert_name, std::move(root_to_report),
+                                    absl::nullopt);
+    }
+    if (info.root_being_watched && root_certificate_.empty()) {
```

Comment:
Noted, thanks. I removed `  && root_certificate_.empty()`.

Reply:
This is incorrect. root_being_watched is not an error.
My earlier question was when does this snippet execute.
Because it seemed like if root_certificate_ is empty it would fail at an earlier point.
Similarly for Key certificate pair

Reply:
I see that you removed the checks. Please do confirm with @ZhenLian, just to make sure we aren't missing anything.